File: python/src/vtx_recon/providers/openai.py
"""OpenAI capability ladder — prove depth of access for a leaked API key.

A TruffleHog ``OpenAI`` finding is a secret API key (``sk-...`` or the project
form ``sk-proj-...``). The ladder is deliberately short because OpenAI exposes
exactly one safe, read-only identity surface and one obviously billable action:

  1. ``list-models``     SAFE. ``GET /v1/models`` — the key authenticates and can
     list the models available to it (read-only, idempotent, non-billable). This
     is the rung that decides VALID vs DENIED.
  2. ``chat-completion`` GATED. ``POST /v1/chat/completions`` — a real completion
     costs the target money. Wrapped with ``@gated``: the safety boundary runs
     *before* any network call, so without BOTH ``--prove`` and an authorized
     scope it raises ``GatedProbeBlocked`` and nothing is sent.

The ladder never raises across its public boundary — every failure becomes a
:class:`ProbeResult`. The raw key is held only transiently for the HTTP call and
is never written into evidence.

AUTHORIZED USE ONLY. See TERMS.md.
"""

from __future__ import annotations

import httpx

from ..models import Finding, LadderResult, ProbeResult, Verdict
from ..redact import redact
from ..safety import Consent, GatedProbeBlocked, ProbeTier, gated
from . import register

__all__ = ["DETECTORS", "openai_ladder"]

# TruffleHog DetectorName values routed to this ladder (matched case-insensitively).
DETECTORS = ("OpenAI",)

_API_BASE = "https://api.openai.com"
_TIMEOUT = httpx.Timeout(10.0)


def _network_failure(name: str, tier: ProbeTier, exc: Exception) -> ProbeResult:
    """Turn an httpx/transport error into a non-success rung (never raise)."""
    return ProbeResult(
        name=name,
        tier=tier,
        success=False,
        detail=f"probe could not complete: {type(exc).__name__}",
        evidence={"error": type(exc).__name__},
    )


def _verdict_from(rungs: list[ProbeResult]) -> Verdict:
    """Derive the impact tier from the rungs that ran.

    * A successful GATED rung that actually ran (not blocked) -> PROVEN.
    * Any successful SAFE rung -> VALID (authenticates + depth shown).
    * The key authenticated nowhere -> DENIED.
    """
    if any(r.success and r.tier is ProbeTier.GATED and not r.blocked for r in rungs):
        return Verdict.PROVEN
    if any(r.success for r in rungs):
        return Verdict.VALID
    return Verdict.DENIED


@register("OpenAI")
async def openai_ladder(finding: Finding, consent: Consent) -> LadderResult:
    """OpenAI ladder: SAFE ``list-models`` -> GATED ``chat-completion``."""
    scope = consent.require_ladder_scope()
    rungs: list[ProbeResult] = []
    key = finding.raw

    models = await _openai_list_models(key)
    rungs.append(models)

    # Ordered: only attempt the gated billable rung if the key authenticated.
    if models.success:
        try:
            rungs.append(await _openai_chat_completion(consent, key))
        except GatedProbeBlocked as blocked:
            rungs.append(
                ProbeResult(
                    name="openai.chat_completion",
                    tier=ProbeTier.GATED,
                    success=False,
                    blocked=True,
                    detail=f"gated billable completion blocked: {blocked.reason}",
                    evidence={"reason": blocked.reason},
                )
            )

    return LadderResult(
        finding=finding,
        provider="openai",
        verdict=_verdict_from(rungs),
        rungs=rungs,
        authorized_scope=scope,
    )


async def _openai_list_models(key: str) -> ProbeResult:
    """SAFE: ``GET /v1/models`` confirms the key and lists reachable models."""
    name = "openai.list_models"
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.get(
                f"{_API_BASE}/v1/models",
                headers={"Authorization": f"Bearer {key}"},
            )
    except httpx.HTTPError as exc:
        return _network_failure(name, ProbeTier.SAFE, exc)

    if resp.status_code != 200:
        return ProbeResult(
            name=name,
            tier=ProbeTier.SAFE,
            success=False,
            detail=f"key rejected (HTTP {resp.status_code})",
            evidence={"status": resp.status_code},
        )

    try:
        body = resp.json()
    except ValueError as exc:
        return _network_failure(name, ProbeTier.SAFE, exc)

    data = body.get("data") if isinstance(body, dict) else None
    models = data if isinstance(data, list) else []
    ids = [m.get("id") for m in models if isinstance(m, dict) and isinstance(m.get("id"), str)]
    return ProbeResult(
        name=name,
        tier=ProbeTier.SAFE,
        success=True,
        detail=f"key authenticates; {len(ids)} models reachable",
        evidence={
            "status": resp.status_code,
            "key_prefix": redact(key),
            "model_count": len(ids),
            "sample_models": ids[:5],
        },
    )


@gated("openai.chat_completion")
async def _openai_chat_completion(_consent: Consent, key: str) -> ProbeResult:
    """GATED: ``POST /v1/chat/completions`` — a billable completion.

    Decorated with ``@gated``: the boundary runs before this body, so without
    BOTH ``--prove`` and an authorized scope it raises ``GatedProbeBlocked`` and
    no billable request is ever sent. A minimal ``max_tokens`` keeps any
    (consented) spend to the smallest possible amount.
    """
    name = "openai.chat_completion"
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
            resp = await client.post(
                f"{_API_BASE}/v1/chat/completions",
                headers={"Authorization": f"Bearer {key}", "Content-Type": "application/json"},
                json={
                    "model": "gpt-4o-mini",
                    "max_tokens": 1,
                    "messages": [{"role": "user", "content": "1"}],
                },
            )
    except httpx.HTTPError as exc:
        return _network_failure(name, ProbeTier.GATED, exc)

    if resp.status_code != 200:
        return ProbeResult(
            name=name,
            tier=ProbeTier.GATED,
            success=False,
            detail=f"billable completion refused (HTTP {resp.status_code})",
            evidence={"status": resp.status_code},
        )

    return ProbeResult(
        name=name,
        tier=ProbeTier.GATED,
        success=True,
        detail="billable chat completion succeeded (spent the target's credits)",
        evidence={"status": resp.status_code, "billable": True},
    )

File: node/src/providers/openai.ts
/**
 * OpenAI capability ladder — prove depth of access for a leaked API key.
 *
 * A TruffleHog `OpenAI` finding is a secret API key (`sk-...` or the project
 * form `sk-proj-...`). The ladder is deliberately short because OpenAI exposes
 * exactly one safe, read-only identity surface and one obviously billable
 * action:
 *
 *   1. `list-models`     `GET /v1/models` — SAFE. The key authenticates and can
 *      list the models available to it (read-only, idempotent, non-billable).
 *      This is what decides VALID vs DENIED.
 *   2. `chat-completion` `POST /v1/chat/completions` — GATED. A real completion
 *      costs the target money. Wrapped with {@link gated}: the safety boundary
 *      runs *before* any network call, so without BOTH `--prove` and an
 *      authorized scope it throws {@link GatedProbeBlocked} and nothing is sent.
 *
 * The ladder never throws across its public boundary — every failure becomes a
 * {@link ProbeResult}. The raw key is held only transiently for the HTTP call
 * and is never written into evidence.
 *
 * AUTHORIZED USE ONLY. See TERMS.md.
 */

import { Finding, LadderResult, ProbeResult, Verdict } from "../models.js";
import { redact } from "../redact.js";
import { Consent, GatedProbeBlocked, ProbeTier, gated } from "../safety.js";
import { HttpError, httpRequest, readJson, type FetchLike } from "./http.js";
import { register } from "./registry.js";

/** TruffleHog DetectorName values routed to this ladder (case-insensitive). */
export const DETECTORS = ["OpenAI"] as const;

const API_BASE = "https://api.openai.com";
const TIMEOUT_MS = 10_000;

function errName(exc: unknown): string {
  if (exc instanceof HttpError) return exc.kind;
  return exc instanceof Error ? exc.constructor.name : "Error";
}

function networkFailure(name: string, tier: ProbeTier, exc: unknown): ProbeResult {
  return new ProbeResult({
    name,
    tier,
    success: false,
    detail: `probe could not complete: ${errName(exc)}`,
    evidence: { error: errName(exc) },
  });
}

/**
 * Derive the impact tier from the rungs that ran.
 * - A successful GATED rung that actually ran (not blocked) -> PROVEN.
 * - Any successful SAFE rung -> VALID (authenticates + depth shown).
 * - The key authenticated nowhere -> DENIED.
 */
function verdictFrom(rungs: ProbeResult[]): Verdict {
  if (rungs.some((r) => r.success && r.tier === ProbeTier.GATED && !r.blocked)) {
    return Verdict.PROVEN;
  }
  if (rungs.some((r) => r.success)) {
    return Verdict.VALID;
  }
  return Verdict.DENIED;
}

/**
 * OpenAI ladder: SAFE `list-models` -> GATED `chat-completion`.
 */
export async function openaiLadder(
  finding: Finding,
  consent: Consent,
  options: { fetchImpl?: FetchLike } = {},
): Promise<LadderResult> {
  const scope = consent.requireLadderScope();
  const rungs: ProbeResult[] = [];
  const key = finding.raw;
  const fetchImpl = options.fetchImpl;

  const models = await openaiListModels(key, fetchImpl);
  rungs.push(models);

  // Ordered: only attempt the gated billable rung if the key authenticated.
  if (models.success) {
    try {
      rungs.push(await openaiChatCompletion(consent, key, fetchImpl));
    } catch (exc) {
      if (exc instanceof GatedProbeBlocked) {
        rungs.push(
          new ProbeResult({
            name: "openai.chat_completion",
            tier: ProbeTier.GATED,
            success: false,
            blocked: true,
            detail: `gated billable completion blocked: ${exc.reason}`,
            evidence: { reason: exc.reason },
          }),
        );
      } else {
        throw exc;
      }
    }
  }

  return new LadderResult({
    finding,
    provider: "openai",
    verdict: verdictFrom(rungs),
    rungs,
    authorizedScope: scope,
  });
}

/** SAFE: `GET /v1/models` confirms the key and lists reachable models. */
async function openaiListModels(key: string, fetchImpl: FetchLike | undefined): Promise<ProbeResult> {
  const name = "openai.list_models";
  let resp: Response;
  try {
    resp = await httpRequest(`${API_BASE}/v1/models`, {
      headers: { Authorization: `Bearer ${key}` },
      timeoutMs: TIMEOUT_MS,
      fetchImpl,
    });
  } catch (exc) {
    return networkFailure(name, ProbeTier.SAFE, exc);
  }

  if (resp.status !== 200) {
    return new ProbeResult({
      name,
      tier: ProbeTier.SAFE,
      success: false,
      detail: `key rejected (HTTP ${resp.status})`,
      evidence: { status: resp.status },
    });
  }

  const body = (await readJson(resp)) as { data?: Array<{ id?: string }> } | undefined;
  if (body === undefined) {
    return networkFailure(name, ProbeTier.SAFE, new SyntaxError("invalid JSON"));
  }

  const models = Array.isArray(body.data) ? body.data : [];
  const ids = models.map((m) => m.id).filter((id): id is string => typeof id === "string");
  return new ProbeResult({
    name,
    tier: ProbeTier.SAFE,
    success: true,
    detail: `key authenticates; ${ids.length} models reachable`,
    evidence: {
      status: resp.status,
      key_prefix: redact(key),
      model_count: ids.length,
      sample_models: ids.slice(0, 5),
    },
  });
}

/**
 * GATED: `POST /v1/chat/completions` — a billable completion.
 *
 * Wrapped with {@link gated}: the boundary runs before this body, so without
 * BOTH `--prove` and an authorized scope it throws {@link GatedProbeBlocked}
 * and no billable request is ever sent. A minimal `max_tokens` keeps any
 * (consented) spend to the smallest possible amount.
 */
export const openaiChatCompletion = gated(
  "openai.chat_completion",
  async (_consent: Consent, key: string, fetchImpl?: FetchLike): Promise<ProbeResult> => {
    const name = "openai.chat_completion";
    let resp: Response;
    try {
      resp = await httpRequest(`${API_BASE}/v1/chat/completions`, {
        method: "POST",
        headers: { Authorization: `Bearer ${key}`, "Content-Type": "application/json" },
        body: JSON.stringify({
          model: "gpt-4o-mini",
          max_tokens: 1,
          messages: [{ role: "user", content: "1" }],
        }),
        timeoutMs: TIMEOUT_MS,
        fetchImpl,
      });
    } catch (exc) {
      return networkFailure(name, ProbeTier.GATED, exc);
    }

    if (resp.status !== 200) {
      return new ProbeResult({
        name,
        tier: ProbeTier.GATED,
        success: false,
        detail: `billable completion refused (HTTP ${resp.status})`,
        evidence: { status: resp.status },
      });
    }

    return new ProbeResult({
      name,
      tier: ProbeTier.GATED,
      success: true,
      detail: "billable chat completion succeeded (spent the target's credits)",
      evidence: { status: resp.status, billable: true },
    });
  },
);

register([...DETECTORS], (finding, consent) => openaiLadder(finding, consent));
